src/libPMacc/include/math/complex/Complex.tpp

```diff
@@ -0,0 +1,203 @@
+{
+    typedef typename ::PMacc::math::Complex<Type> result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<Type>& other)
+    {
+        return PMacc::algorithms::math::euler(1.0,other.get_imag())*PMacc::algorithms::math::exp(other.get_real());
+    }
+};
+
+/*  Set primary template and subsequent specialization of arg() for retrieving
+ *  the phase of a complex number (Note: Branchcut running from -infinity to 0).
+ */
+template<typename Type>
+struct Arg;
+
+template<typename Type>
+HDINLINE typename Arg< Type >::result arg(const Type& val)
+{
+    return Arg< Type > ()(val);
+}
+
+template<typename Type>
+struct Arg< ::PMacc::math::Complex<Type> >
+{
+    typedef typename ::PMacc::math::Complex<Type>::type result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<Type>& other)
+    {
+        if (other.get_real()==0.0 && other.get_imag()==0.0) return 0.0;
+        else if (other.get_real()==0.0 && other.get_imag()>0.0) return M_PI/2;
+        else if (other.get_real()==0.0 && other.get_imag()<0.0) return -M_PI/2;
+        else if (other.get_real()<0.0 && other.get_imag()==0.0) return M_PI;
+        else return PMacc::algorithms::math::atan2(other.get_imag(),other.get_real());
+    }
+};
+
+/*  Specialize pow() for complex numbers. */
+template<typename Type>
+struct Pow< ::PMacc::math::Complex<Type>, Type >
+{
+    typedef typename ::PMacc::math::Complex<Type> result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<Type>& other, const Type& exponent)
+    {
+        return PMacc::algorithms::math::pow( PMacc::algorithms::math::abs(other),exponent )
+                *PMacc::algorithms::math::exp( ::PMacc::math::Complex<Type>(0.,1.)*PMacc::algorithms::math::arg(other)*exponent );
+    }
+};
+
+/*  Specialize abs() for complex numbers. */
+template<typename Type>
+struct Abs< ::PMacc::math::Complex<Type> >
+{
+    typedef typename ::PMacc::math::Complex<Type>::type result;
+
+    HDINLINE result operator( )(const ::PMacc::math::Complex<Type>& other)
+    {
+        return PMacc::algorithms::math::sqrt( PMacc::algorithms::math::abs2(other.get_real()) + PMacc::algorithms::math::abs2(other.get_imag()) );
+    }
+};
+
+/*  Specialize abs2() for complex numbers. */
+template<typename Type>
+struct Abs2< ::PMacc::math::Complex<Type> >
+{
+    typedef typename ::PMacc::math::Complex<Type>::type result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<Type>& other)
+    {
+        return PMacc::algorithms::math::abs2(other.get_real()) + PMacc::algorithms::math::abs2(other.get_imag());
+    }
+};
+
+} //namespace math
+} //namespace algorithms
+} // namespace PMacc
+
+namespace PMacc
+{
+namespace algorithms
+{
+namespace precisionCast
+{
+
+/*  Specialize precisionCast-operators for complex numbers. */
+
+template<typename CastToType>
+struct TypeCast<CastToType, ::PMacc::math::Complex<CastToType> >
+{
+    typedef const ::PMacc::math::Complex<CastToType>& result;
```

Comment:
Yes by `const ref` is right. 
If the type has already the right precision there is no need to copy.